posix/devices/hostsock/hostsock.c

```diff
@@ -0,0 +1,957 @@
+    oe_fd_t* sock_,
+    struct oe_sockaddr* addr,
+    oe_socklen_t* addrlen)
+{
+    int ret = -1;
+    sock_t* sock = _cast_sock(sock_);
+    oe_socklen_t addrlen_in = 0;
+
+    oe_errno = 0;
+
+    if (!sock)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (addrlen)
+        addrlen_in = *addrlen;
+
+    if (oe_posix_getsockname_ocall(
+            &ret,
+            sock->host_fd,
+            (struct oe_sockaddr*)addr,
+            addrlen_in,
+            addrlen) != OE_OK)
+    {
+        OE_RAISE_ERRNO(OE_EINVAL);
+    }
+
+done:
+
+    return ret;
+}
+
+static ssize_t _hostsock_read(oe_fd_t* sock_, void* buf, size_t count)
+{
+    return _hostsock_recv(sock_, buf, count, 0);
+}
+
+static ssize_t _hostsock_write(oe_fd_t* sock_, const void* buf, size_t count)
+{
+    return _hostsock_send(sock_, buf, count, 0);
+}
+
+static ssize_t _hostsock_readv(
+    oe_fd_t* desc,
+    const struct oe_iovec* iov,
+    int iovcnt)
+{
+    ssize_t ret = -1;
+    sock_t* sock = _cast_sock(desc);
+    void* buf = NULL;
+    size_t buf_size = 0;
+
+    if (!sock || (!iov && iovcnt) || iovcnt < 0 || iovcnt > OE_IOV_MAX)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* Flatten the IO vector into contiguous heap memory. */
+    if (oe_iov_pack(iov, iovcnt, &buf, &buf_size) != 0)
+        OE_RAISE_ERRNO(OE_ENOMEM);
+
+    /* Call the host. */
+    if (oe_posix_recvv_ocall(&ret, sock->host_fd, buf, iovcnt, buf_size) !=
+        OE_OK)
+    {
+        OE_RAISE_ERRNO(OE_EINVAL);
+    }
+
+    /* Synchronize data read with IO vector. */
+    if (oe_iov_sync(iov, iovcnt, buf, buf_size) != 0)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+done:
+
+    if (buf)
+        oe_free(buf);
+
+    return ret;
+}
+
+static ssize_t _hostsock_writev(
+    oe_fd_t* desc,
+    const struct oe_iovec* iov,
+    int iovcnt)
+{
+    ssize_t ret = -1;
+    sock_t* sock = _cast_sock(desc);
+    void* buf = NULL;
+    size_t buf_size = 0;
+
+    if (!sock || !iov || iovcnt < 0 || iovcnt > OE_IOV_MAX)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* Flatten the IO vector into contiguous heap memory. */
+    if (oe_iov_pack(iov, iovcnt, &buf, &buf_size) != 0)
+        OE_RAISE_ERRNO(OE_ENOMEM);
+
+    /* Call the host. */
+    if (oe_posix_sendv_ocall(&ret, sock->host_fd, buf, iovcnt, buf_size) !=
+        OE_OK)
+    {
+        OE_RAISE_ERRNO(OE_EINVAL);
+    }
+
+done:
+
+    if (buf)
+        oe_free(buf);
+
+    return ret;
+}
+
+static int _hostsock_socket_shutdown(oe_fd_t* sock_, int how)
+{
+    int ret = -1;
+    sock_t* sock = _cast_sock(sock_);
+
+    oe_errno = 0;
+
+    if (!sock)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (oe_posix_shutdown_ocall(&ret, sock->host_fd, how) != OE_OK)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+done:
+
+    return ret;
+}
+
+static int _hostsock_release(oe_device_t* device_)
+{
+    int ret = -1;
+    device_t* device = _cast_device(device_);
+
+    oe_errno = 0;
+
+    if (!device)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    ret = 0;
```

Comment:
I might be missing something, but nothing is actually released here.

Reply:
It looks like it was actually freed in the socket close call, which makes sense to me.

Reply:
This release function is for the device that creates sockets, not the sockets themselves. This particular devices is static and has no allocated storage. The device->release() method is for cleaning up any memory when a device is removed from the device table. In this case there is none. I added a comment to that effect.